### DISCUSSION TITLE
Prediction on CPU

### DISCUSSION THREAD

**Discussion by Amalkas** [Abe404/root_painter#117] [Q&A] [answered]
Hi,

First of all, thank you for the super cool tool! RootPainter has saved me hours of work.

I have paid for Google Colab and managed to train a quite decent model for my data.

Is it possible now to run RootPainter server on my local machine using CPU for calculations? Prediction should be much less demanding than training, so I think my CPU should be able to manage with reasonable performance (even several minutes per image would be still fine).

**Comment by Abe404** [comment]
Hi! I'm glad you have found RootPainter useful!

I will need to make some modifications to the code, but it should be possible. I will look into this tomorrow and let you know how it goes.

For most use cases, I think colab would still make more sense, as I expect the GPU segmentation to be much faster, but I still think it makes sense to enable CPU inference to provide more options.
I will update you soon.

Kind regards,
Abraham

**Reply by Amalkas** [reply]
Hi,
Thanks! I also have a CUDA-compatible GPU, but it's quite old (GT 710). There's no pre-compiled PyTorch version available for it. Frankly, I have no idea what would be more meaningful: compiling Torch with ancient CUDA or trying to run the RP server on CPU. Unfortunately, Colab is not an option.
Best

**Comment by Abe404** [comment, marked as answer]
Hi,

I have updated the code (on the master branch) to handle CPU inference.

The changes are visible here: https://github.com/Abe404/root_painter/compare/78eda63ccf589241b988bba20e68bccab2f96390..6c4f7c8c5d18905bc62455e74f1924029d27a50b (also including an update to handle more recent versions of PIL image library).

I tested with images with width of 762 and height of 392 (pixels) and it took less than 5 seconds per image for inference.

Much larger images will take much longer.

Please let me know if this solves the problem for you. 

Working with this updated code requires installing the server by cloning/downloading the git repo. Let me know if you run into trouble with that.

Kind regards,
Abraham

**Reply by Amalkas** [reply]
Cool! It takes 15 sec in my case for 800x800 images, and it's actually far beyond my expectations !
Thank you a lot!

**Comment by Abe404** [comment]
Super! I'm happy we were able to find a solution. 

Let me know if you run into any more problems. I'd also be interested to learn more about your project, including the type of data you are working with and how RootPainter has helped you, if/when you have anything you can share (ideally images). It's useful to know how RootPainter performs on different datasets so we can make suggestions for other users.

Kind regards,
Abraham
